File: 0503-next-greater-element-ii/0503-next-greater-element-ii.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        if nums is None or len(nums) == 0:
            return [-1]

        n = len(nums)  #len of nums array
        result = [-1 for i in range(n)]
        st = [] #decreasing monotonic stack to index of the elements

        for index in range((n * 2)):
                
            currIndex = index % (n)
            # print(st, currIndex, nums[currIndex])
            # print(st, nums[currIndex] ,nums[st[-1]])
            while st and nums[currIndex] > nums[st[-1]]:
                top = st.pop(-1)
                result[top] = nums[currIndex]

            if index < n:
                st.append(index)
        
        return result

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        if(nums == null || nums.length == 0){
            return new int[-1];
        }

        int n = nums.length;
        Stack<Integer> st = new Stack<>();
        int[] result = new int[n];

        Arrays.fill(result, -1);

        for(int index = 0; index < (n * 2); index++){

            int currIndex = index % n;

            while(!st.isEmpty() && nums[currIndex] > nums[st.peek()]){
                int top = st.pop();
                result[top] = nums[currIndex];
            }

            if(index < n){
                st.add(currIndex);
            }
        }
        return result;
    }
}

// #monotonic stack
// #Time Complexity: O(n)
// #Space Complexity:O(n)